File: Design Patterns/Creational/Abstract factory/Delphi/Factory.pas
unit Factory;

interface

uses
  SysUtils, Classes;

type
  IButton = interface
    procedure Render;
  end;

  IFactory = interface
    function CreateButton(): IButton;
  end;

  TWinButton = class(TInterfacedObject, IButton)
    procedure Render;
  end;

  TMacButton = class(TInterfacedObject, IButton)
    procedure Render;
  end;

  TWinFactory = class(TInterfacedObject, IFactory)
    function CreateButton(): IButton;
  end;

  TMacFactory = class(TInterfacedObject, IFactory)
    function CreateButton(): IButton;
  end;

implementation

procedure TWinButton.Render;
begin
  Writeln('Drawing the Windows button');
end;

procedure TMacButton.Render;
begin
  Writeln('Drawing the macOS button');
end;

function TWinFactory.CreateButton: IButton;
begin
  Result := TWinButton.Create;
end;

function TMacFactory.CreateButton: IButton;
begin
  Result := TMacButton.Create;
end;

end.

File: Creational/Abstract factory/C#/AbstractFactoryExample/AbstractFactoryExample/Factory.cs
namespace AbstractFactoryExample;

public interface IGuiFactory
{
	IButton CreateButton();
}

public interface IButton
{
	void Render();
}

public class WindowsButton : IButton
{
	public void Render() => Console.WriteLine("Drawing a Windows window");
}

public class MacButton : IButton
{
	public void Render() => Console.WriteLine("Drawing a macOS window");
}

public class WindowsGuiFactory : IGuiFactory
{
	public IButton CreateButton() => new WindowsButton();
}

public class MacGuiFactory : IGuiFactory
{
	public IButton CreateButton() => new MacButton();
}